Community/Expire/automation/jsr223/python/community/expire/expire.py

```diff
@@ -103,127 +106,239 @@ def get_config(item_name):
         - item: The Item that expired.
-        - cfg: Contians a dict representation of the expire config returned by
-        get_config.
-        - log: Logger from the expire Rule.
+        - exp_type: The action type, 'state' or 'command'.
+        - exp_state: The state to expire to.
+        - log: Logger from the expire rule.
     """
-    log.debug("{} expired, {} to {}".format(item.name, exp_type, exp_state))
-
-    # Force the state to a StringType for StringItems to allow us to set the
-    # Item to "UNDEF" and "NULL" as opposed to the UnDefTypes.
-    if item.type == "String" and isinstance(exp_state, basestring):
-        exp_state = StringType(exp_state)
+    log.debug("'{}' expired, {} '{}'".format(item.name, exp_type, exp_state))
 
     if exp_type == "state":
         events.postUpdate(item, exp_state)
-    else:
+    elif exp_type == "command":
         events.sendCommand(item, exp_state)
+    else:
+        log.warn("Unrecognized action '{}' for item '{}'"
+                 .format(exp_type, item.name))
+
 
-@rule("Expire",
-      description=("Simulates the Expire1 binding, updating or commanding an "
-                   "Item after a comnfigured amount of time."),
-       tags=["expire"])
-@trigger_generator()
 def expire(event):
     """
-    Called when a member of the configured Group changes. The name of the Group
-    is imported from configuration as expire_items_gr.
+    Called when an item configured for expire receives an update.
 
     If the Item updates to an UnDefType, the change is ignored.
     If the Item updates to the same state configured in the expire config any
     expire timer is cancelled.
     If the Item updates to a different state configured in the expire config a
     new timer is created to go off at the configured time in the future.
     """
+    item_name = event.itemName
+
     # Ignore changes to an UnDefType.
     if isinstance(event.itemState, UnDefType):
+        if (timers[item_name] is not None
+                and not timers[item_name].hasTerminated()):
+            timers[item_name].cancel()
         return
 
-    cfg = get_config(event.itemName)
+    cfg = get_config(item_name, expire.log)
+    if not cfg:
+        expire.log.warn("Skipping expire processing for '{}'"
+                        .format(item_name))
+        if (timers[item_name] is not None
+                and not timers[item_name].hasTerminated()):
+            timers[item_name].cancel()
+        return
 
     # Cancel the timer when the Item enters the cfg state.
     # Use unicode because there is no degree symbol in ASCII so we can handle
     # Number:Temperature Items.
-    if unicode(items[event.itemName]) == cfg["state"]:
-        if (event.itemName in timers
-                and not timers[event.itemName].hasTerminated()):
-            timers[event.itemName].cancel()
-            del timers[event.itemName]
+    if unicode(items[item_name]) == cfg["state"]:
+        if (timers[item_name] is not None
```

Comment:
If you try to access a value from a Dict that doesn't exist you get a KeyError. That's why I use the `in` operator. The above code (and other places where you replaced the `in` with `is not None` will throw and exception. At least it did for me and that appears to the expected behavior.

Reply:
I've changed it to create an entry in the dictionary for every configured item. This is used to track removed items during a reload.